src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -270,6 +270,152 @@ void ucc_tl_mlx5_mcast_allgather_progress(ucc_coll_task_t *coll_task)
+                                                      ucc_tl_mlx5_mcast_coll_req_t  *req)
+{
+    if ((req->concurreny_level % 2 == 0 && req->num_packets % req->mcast_prepost_bucket_size != 0) ||
+        (comm->commsize % req->concurreny_level != 0) ||
+        (req->length % comm->max_per_packet != 0)) {
+        tl_warn(comm->lib, "Pipelined mcast allgather not supported: "
+                "num_packets %d mcast_prepost_bucket_size %d "
+                "length %ld max_per_packet %d "
+                "team size %d concurreny_level %d",
+                req->num_packets, req->mcast_prepost_bucket_size, req->length,
+                comm->max_per_packet, comm->commsize, req->concurreny_level);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    if (req->mcast_prepost_bucket_size * req->concurreny_level * 2 > comm->params.rx_depth) {
+        tl_warn(comm->lib, "Pipelined mcast allgather not supported: "
+                "either reduce prepost_bucket_size or mcast group "
+                "count or increase recv queue size "
+                "mcast_prepost_bucket_size %d concurreny_level %d "
+                "rx_depth %d",
+                 req->mcast_prepost_bucket_size, req->concurreny_level,
+                 comm->params.rx_depth);
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    return UCC_OK;
+}
+
+
+/*
+ * at each stage half of the mcast groups are ready for receiving mcast
+ * packets while the other half are getting prepared by preposting recv
+ * buffers
+ */
+static inline ucc_status_t
+ucc_tl_mlx5_mcast_prepare_zero_copy_allgather(ucc_tl_mlx5_mcast_coll_comm_t *comm,
+                                              ucc_tl_mlx5_mcast_coll_req_t  *req)
+{
+    ucc_tl_mlx5_mcast_reg_t                   *reg    = NULL;
+    ucc_rank_t                                 root   = 0;
+    int                                        offset = 0;
+    ucc_status_t                               status;
+    ucc_rank_t                                 j, i;
+    int                                        total_steps;
+    ucc_tl_mlx5_mcast_pipelined_ag_schedule_t *new_sched;
+
+    ucc_assert(comm->allgather_comm.truly_zero_copy_allgather_enabled);
+
+    req->concurreny_level = comm->mcast_group_count / 2;
+    req->concurreny_level = ucc_min(req->concurreny_level, ONE_SIDED_MAX_CONCURRENT_LEVEL);
+    req->concurreny_level = ucc_min(req->concurreny_level, comm->commsize);
+
+    if (req->concurreny_level == 0) {
+        tl_warn(comm->lib, "not enough concurreny level to enable zcopy pipeline allgather");
+        return UCC_ERR_NOT_SUPPORTED;
+    }
+
+    if (comm->allgather_comm.mcast_prepost_bucket_size > req->num_packets) {
+        req->mcast_prepost_bucket_size = req->num_packets;
+    } else {
+        req->mcast_prepost_bucket_size = comm->allgather_comm.mcast_prepost_bucket_size;
+    }
+
+	status = ucc_tl_mlx5_mcast_validate_zero_copy_allgather_params(comm, req);
+	if (status != UCC_OK) {
+        return status;
+    }
```

Comment:
indent issue

Reply:
oh sorry I missed the line which had issue. It is fixed now.